### ISSUE TITLE
New Rule request: Forbidden API calls

### ISSUE THREAD

**Issue by bmhm** [checkstyle/checkstyle#7449]
Check documentation: no such check yet, I checked.

```

/var/tmp $ cat config.xml
<module name=ForbiddenAPICalls" />

/var/tmp $ cat YOUR_FILE.java
Optional<String> optStr = Optional.of("");
optStr.get(); 
```

---------------

Describe what you expect in detail.
Line 2 should throw a Checkstyle error (configurable rule violation).

Rationale:
Brian Goetz mentioned on a few occasions that the method is superfluous and has a misleading name.

Others want to forbid specific API calls of other libraries. That said, Optional.get() seems like a sensible default, but more methods should be configurable.

--------------

**Comment by romani** [comment]
checkstyle is not type aware tool - https://checkstyle.org/writingchecks.html#Limitations
so we cannot reliably distinguish classes and types of arguments.
If you ready to provide some implementation that will work in some/most cases , we can host it in sevntu project 

You can see similar implementation http://sevntu-checkstyle.github.io/sevntu.checkstyle/apidocs/com/github/sevntu/checkstyle/checks/coding/ForbidCertainMethodCheck.html , https://github.com/sevntu-checkstyle/sevntu.checkstyle/issues/604 , problems of it https://github.com/sevntu-checkstyle/sevntu.checkstyle/issues/773 .

**Comment by bmhm** [comment]
Hello @romani, thanks for the info. Maybe this is a task for another tool, then? I am thinking of tools like Google's spotbugs, Google's error-prone or similar tools. Maybe even PMD?

**Comment by romani** [comment]
I think better location is https://github.com/policeman-tools/forbidden-apis we use this tool in our build

**Comment by bmhm** [comment]
Thanks! That's a case for: using the right tool for the right job.
